### PR TITLE
Readme updates

### PR DESCRIPTION
**Pull request by agubler**
**Type:** feature

The following has been addressed in the PR:

* [ ] There is a related issue
* [ ] All code matches the [style guide](https://github.com/dojo/meta/blob/master/STYLE.md)
* [ ] Unit or Functional tests are included in the PR

**Description:**

README.md updates.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -117,60 +117,61 @@ The following code creates an element with the specified `tag`
 v(tag: string): HNode[];
 ```
 
-where `tag` is in the form: element.className(s)#id, e.g.
```
Comment (rishson) [inline README.md]: Is this no longer supported?

File: README.md [hunk 2]
```diff
@@ -479,6 +509,89 @@ registry.define('my-widget', () => {
 });
 ```
 
+#### Injecting State
+
+Working with larger widget structures, it can quickly become tiresome and complex to pass all the required properties down the tree. Additionally having to do so often means that widgets are forced to be aware of properties that they they don't need, but are required to know about in order to ensure they are propagated to the children.
```
Comment (rishson) [inline README.md]: I think we need a section upfront that states all widgets are linked in a tree. I know its obvious, but it should still be stated.

File: README.md [hunk 3]
```diff
@@ -479,6 +509,89 @@ registry.define('my-widget', () => {
...
+
+Dojo 2 provides a mechanism to inject state directly to parts of the widget structure; this is done by defining an `Injector` in the `registry` and passing a context object that will source the state that is to be injected.
```
Comment (rishson) [inline README.md]: 'widget structure' -> widget tree, tree of widgets. Don't mind, but needs to be consistent with section above.

File: README.md (hunk 1 above)
Comment (agubler) [inline README.md]: We don't use it in any of our widgets and it would not be considered a "best practice" to. So I think it's best to not remove it from the readme.

File: README.md [hunk 4]
```diff
@@ -479,6 +509,89 @@ registry.define('my-widget', () => {
...
+import { Injector, BaseInjector } from '@dojo/widget-core/Injector';
+
+const myStateContext = {
+	theme: 'solid'
+};
+
+registry.define('state', Injector(BaseInjector, myStateContext));
+```
+
+To use the injected state, create a `beforeRender` method that returns a render function which creates a `w` reference to the registry item like any other widget and pass the properties required by the `InjectorProperties` interface.
```
Comment (rishson) [inline README.md]: Maybe add an explanation of the above code, e.g. Here you can see we created a context called `myStateContext` and added a single state property `theme` to it, a gave this property the value of `solid`? or too tutorially?

File: README.md [hunk 5]
```diff
@@ -479,6 +509,89 @@ registry.define('my-widget', () => {
...
+			getChildren(context: any, children: DNode[]): DNode[] {
+				return [];
+			},
+			children
+		});
+	};
+}
+```
+
+This will inject the values of `myState` as properties to the widget, as the returned object from `getProperties` is mixed over the widgets' existing properties.
```
Comment (rishson) [inline README.md]: I don't see how the above code will inject state. Can we say how, even if its hidden and automatic?

File: README.md [hunk 6]
```diff
@@ -479,6 +509,89 @@ registry.define('my-widget', () => {
...
+			},
+			children
+		});
+	};
+}
+```
+
+This will inject the values of `myState` as properties to the widget, as the returned object from `getProperties` is mixed over the widgets' existing properties. 
+
+For convenience, Dojo 2 provides mixin called `Container` that will decorate a widget with the above `beforeRender` implementation. Using the `Container` mixin enables any view widget to have state injected without coupling the widget to always have state injected. This means the widget can also be used as a normal widget with properties being passed from it's parent.
```
Comment (rishson) [inline README.md]: view widget is a term that needs defining, e.g. aren't all widgets 'view widgets' but some can also be container widgets?

File: README.md [hunk 7]
```diff
@@ -479,6 +509,89 @@ registry.define('my-widget', () => {
...
+import { Container } from '@dojo/widget-core/Container';
+
+function getProperties(context: any, properties: any): any {
+	return context;
+}
+
+const MyViewWidgetContainer = Container(MyViewWidget, 'state', { getProperties });
+```
+
+**Note:** that both the `getProperties` and `getChildren` functions do not need to be provided, if the functions are not defined the default mappers will be used that return an empty object and an empty array respectively.
```
Comment (rishson) [inline README.md]: Should we have a section 'Default container implementation' that doesn't have the `getProperties` function, and then a section on 'Advanced container implementation' that is like the code below?

File: README.md (hunk 2 above)
Comment (agubler) [inline README.md]: Perhaps, but `injecting state` is not the place for that (I don't know where that would actually go in the current structure).

File: README.md (hunk 5 above)
Comment (agubler) [inline README.md]: The state will get injected as properties, as the returned object from `getProperties` is mixed over the existing `properties`, is that not enough to explain how?

File: README.md (hunk 6 above)
Comment (agubler) [inline README.md]: Well view widgets is a pretty common term for a component that only has presentational concerns; I was trying to emphasise that if a consumer implements the `beforeRender` manually then they have coupled the widget to having state injected.

The `Container` is a way that a consumer can connect the injected state to an existing widgets without having to modify or couple the widget to having state injected.

For example we could take `Checkbox` from `@dojo/widgets` and "connect" that to some state but also use the `Checkbox` elsewhere as a "normal" Dojo 2 widget.

File: README.md (hunk 7 above)
Comment (agubler) [inline README.md]: I don't think so, the default implementation is effectively a `noop`. Just wanted to indicate that you don't need to provide both mappers, if you only want to inject properties.

File: README.md (hunk 4 above)
Comment (agubler) [inline README.md]: Perhaps a bit tutorially, happy to add something if there is a consensus.

File: README.md (hunk 1 above)
Comment (rishson) [inline README.md]: 👍 If its still _possible_ though, then maybe add to the reference guide (when written).

File: README.md (hunk 2 above)
Comment (rishson) [inline README.md]: Maybe in the `w` section? Something like:

```ts
[in Baz.ts]

protected render: DNode {
  return v(div, {}, [
    w(Foo, {}, [Qux]);
    w(Bar, {});
  ]);
}
```
So now you have created instances of `Foo` and `Bar` widgets from inside the `render` call of `Baz` widget. Because you passed a child `Qux` to the `w` call of `Foo`, you have created a tree of widgets like so:

```
div (Baz)
  |
  - Foo
  |   |
  |   - Qux
  |
  - Bar
```
Note that in the DOM, you will see the result of the `render` call for `Foo, Bar, and Qux`;

Comment (agubler) [inline README.md]: Do we really think we need to explain this in the technical readme?

Comment (rishson) [inline README.md]: No - OK. This should be added to the beta2 tutorial 'State containers and advanced property passing'.

Comment (agubler) [inline README.md]: I would imagine the fact that the widgets are a tree would be in one of the the starter tutorials?

File: README.md [hunk 8]
```diff
@@ -303,7 +295,7 @@ class MyWidget extends WidgetBase<WidgetProperties> {
 }
 ```
 
-*registering the `onPropertiesChanged ` function in the constructor*
+For non decorator enviroments the listener can be registered using the `onPropertiesChanged ` function in the constructor.
```
Comment (dylans) [inline README.md]: enviroments -> environments

File: README.md [hunk 9]
```diff
@@ -336,11 +328,51 @@ Finally once all the attached events have been processed, the properties lifecyc
 
 <!-- TODO: render lifecycle goes here -->
 
-##### AfterRender
+Occassionally, in a mixin or base widget class, it my be required to provide logic that needs to be executed before or after a widgets' `render` call. These lifecycle hooks are supported in `WidgetBase` and operate as before and after aspects.
```
Comment (dylans) [inline README.md]: widgets' -> widget's (unless you meant widgets to be both plural and possessive)

Comment (dylans) [inline README.md]: Is it clear in this context what before and after aspects are? Should we link to something about AOP style advice or ?

File: README.md [hunk 10]
```diff
@@ -336,11 +328,51 @@ Finally once all the attached events have been processed, the properties lifecyc
...
-##### AfterRender
+Occassionally, in a mixin or base widget class, it my be required to provide logic that needs to be executed before or after a widgets' `render` call. These lifecycle hooks are supported in `WidgetBase` and operate as before and after aspects. 
+
+The functionality is provided by the `beforeRender` and `afterRender` decorators.
+
+***Note:*** Both the `beforeRender` and `afterRender` functions are executed in the order that they are specified from the super class up to the final class.
+
+##### BeforeRender
+
+The `beforeRender` call receives the widgets' `render` function, `properties` and `children` and is expected to return a function that satifies the `render` API. The `properties` and `children` are passed to enable them to be manipulated or decorated prior to the `render` being called.
```
Comment (dylans) [inline README.md]: widgets' -> widget's

File: README.md [hunk 11]
```diff
@@ -336,11 +328,51 @@ Finally once all the attached events have been processed, the properties lifecyc
...
+
+The functionality is provided by the `beforeRender` and `afterRender` decorators.
+
+***Note:*** Both the `beforeRender` and `afterRender` functions are executed in the order that they are specified from the super class up to the final class.
+
+##### BeforeRender
+
+The `beforeRender` call receives the widgets' `render` function, `properties` and `children` and is expected to return a function that satifies the `render` API. The `properties` and `children` are passed to enable them to be manipulated or decorated prior to the `render` being called. 
+
+This is the only time in the widget lifecycle that exposes either of these attributes to be manipluted outside of the property system.
```
Comment (dylans) [inline README.md]: Maybe add a sentence or two on why, and why this is a good thing?

File: README.md [hunk 12]
```diff
@@ -336,11 +328,51 @@ Finally once all the attached events have been processed, the properties lifecyc
...
+		properties.extraAttribute = 'foo';
+		// Return or replace the `render` function
+		return () => {
+			return v('my-replaced-attribute');
+		};
+	}
+}
+```
+
+And using the beforeRender function for non decorator environments:
```
Comment (dylans) [inline README.md]: `beforeRender`

File: README.md [hunk 13]
```diff
@@ -336,11 +328,51 @@ Finally once all the attached events have been processed, the properties lifecyc
...
+		return renderFunc;
+	}
+}
+```
 
-Occassionally, in a mixin or base widget class, it my be required to provide logic that needs to be executed using the result of a widget's `render`. `WidgetBase` supports registering functions that opererate as an after aspect to the `render` function using a provided decorator `afterRender` or in non-decorator environments by calling the decorator method directly.
+##### AfterRender
 
-*Using the `afterRender` decorator*
+The `afterRender` call receives the returned `DNode`s from a widgets' `render` call, so that the nodes can decorated, manipulated or even swapped.
```
Comment (dylans) [inline README.md]: widgets' -> widget's

File: README.md [hunk 14]
```diff
@@ -352,7 +384,7 @@ class MyBaseClass extends WidgetBase<WidgetProperties> {
 }
 ```
 
-*Without using decorators*
+And using the afterRender function for non decorator environments:
```
Comment (dylans) [inline README.md]: `afterRender`

File: README.md (hunk 2 above)
Comment (dylans) [inline README.md]: "Additionally having to do so often means that widgets are forced to be aware of properties that they they don't need, but are required to know about in order to ensure they are propagated to the children." is a bit hard to follow, perhaps something like "Needing to frequently pass all required properties also requires widgets to be aware of properties that are only needed for ensuring their propagation to child widgets."

File: README.md [hunk 15]
```diff
@@ -479,6 +509,89 @@ registry.define('my-widget', () => {
...
+			getChildren(context: any, children: DNode[]): DNode[] {
+				return [];
+			},
+			children
+		});
+	};
+}
+```
+
+This will inject the values of `myState` as properties to the widget, as the returned object from `getProperties` is mixed over the widgets' existing properties.
```
Comment (dylans) [inline README.md]: widgets' -> widget's

File: README.md [hunk 16]
```diff
@@ -479,6 +509,89 @@ registry.define('my-widget', () => {
...
+			},
+			children
+		});
+	};
+}
+```
+
+This will inject the values of `myState` as properties to the widget, as the returned object from `getProperties` is mixed over the widgets' existing properties. 
+
+For convenience, Dojo 2 provides mixin called `Container` that will decorate a widget with the above `beforeRender` implementation. Using the `Container` mixin enables any view widget to have state injected without coupling the widget to always have state injected. This means the widget can also be used as a normal widget with properties being passed from it's parent.
```
Comment (dylans) [inline README.md]: "Dojo 2 provides mixin called `Container`" -> "Dojo 2 provides a mixin called `Container`"

Comment (dylans) [inline README.md]: it's -> its

File: README.md [hunk 17]
```diff
@@ -479,6 +509,89 @@ registry.define('my-widget', () => {
...
+function getProperties(context: any, properties: any): any {
+	return context;
+}
+
+const MyViewWidgetContainer = Container(MyViewWidget, 'state', { getProperties });
+```
+
+**Note:** that both the `getProperties` and `getChildren` functions do not need to be provided, if the functions are not defined the default mappers will be used that return an empty object and an empty array respectively.
+
+There may be times when the default `BaseInjector` doesn't meet the all your needs. For example if the context contains a reference to an eventable instance, you may want to add an event listener in the Injector to perform some logic, perhaps invalidate the widget.
```
Comment (dylans) [inline README.md]: "doesn't meet the all your needs." -> doesn't fully meet your needs.""

Comment (dylans) [inline README.md]: Injector -> `Injector` ?

File: README.md [hunk 18]
```diff
@@ -852,15 +965,15 @@ to the widget instance.
 
 These are some of the **important** principles to keep in mind when creating and using widgets:
 
-1. The widget *`__render__`* function should **never** be overridden
+1. The widgets' *`__render__`*, *`__setProperties__`*, *`__setChildren__`* functions should **never** be called or overridden
```
Comment (dylans) [inline README.md]: widgets' -> widget's

**Comment by agubler** [comment]
@rishson @dylans I believe i've addressed or responded to all the feedback :)
